File: PFinalQytetetRuby/QytetetRuby/lib/jugador.rb
#encoding: UTF-8



module ModeloQytetet
  class Jugador 
       
    attr_accessor :encarcelado, :cartaLibertad, :casillaActual
    attr_reader :nombre, :saldo, :propiedades    
    
    
    def initialize(nombre, cartaLibertad=nil, casillaActual=nil, propiedades=Array.new, encarcelado = false, saldo = 7500 ) 
        @nombre = nombre
        @cartaLibertad = cartaLibertad
        @casillaActual = casillaActual
        @propiedades = propiedades
        @encarcelado = encarcelado
        @saldo = saldo    
    end
    
    
    def cancelarHipoteca(titulo)
      cancelar = false
      costecancelar = titulo.calcularCosteCancelar()
        
      if (tengoSaldo(costecancelar))
            
          modificarSaldo(-costecancelar)
          titulo.cancelarHipoteca()
          cancelar = true
      end        
        
      return cancelar
    end
     
    def comprarTituloPropiedad()
      comprado = false
        
      costeCompra = @casillaActual.coste
      if (costeCompra < @saldo)
          titulo = @casillaActual.asignarPropietario(self) #jugadorActual          
          @propiedades << (titulo)
          modificarSaldo(-costeCompra)
            
          comprado = true
            
      end
      return comprado
    end
     
    def cuantasCasasHotelesTengo()
      num = 0
      @propiedades.each do |prop|      
          num += prop.numCasas + prop.numHoteles
      end
      return num
    end
     
    def deboPagarAlquiler()
      titulo = @casillaActual.titulo
        
      esDeMiPropiedad = esDeMiPropiedad(titulo)
      tienePropietario = false
      encarcelado = true
      estaHipotecada = true
        
        
        
      if (!esDeMiPropiedad)
          tienePropietario =  titulo.tengoPropietario()
          
          if(tienePropietario)
              encarcelado = titulo.propietarioEncarcelado()
          
              estaHipotecada = titulo.hipotecada
          end        
      end
                    
      return !esDeMiPropiedad && tienePropietario && !encarcelado && !estaHipotecada 
    end
     
    def devolverCartaLibertad()
      cartaLibertad = @cartaLibertad
      @cartaLibertad = nil
      return cartaLibertad
    end
     
    def edificarCasa(titulo)
      return puedoEdificarCasa(titulo)
    end
     
    def edificarHotel(titulo)
      return puedoEdificarHotel(titulo)
    end
     
     
    def eliminarDeMisPropiedades(titulo)
      @propiedades.delete(titulo)
      titulo.propietario=(nil)
    end
     
    def esDeMiPropiedad(titulo)
      return @propiedades.include?(titulo)
    end
     
    def estoyEnCalleLibre()
      raise NotImplementedError
    end
     
    
    def hipotecarPropiedad(titulo)
      costeHipoteca = titulo.hipotecar()
      modificarSaldo(costeHipoteca)
    end
    
    def irACarcel(casilla)
      @casillaActual=(casilla)
      @encarcelado=(true)
    end
    
    def modificarSaldo(cantidad)
      return @saldo += cantidad  
    end
    
    def obtenerCapital()
      capital = @saldo
      valorProp = 0
        
      @propiedades.each do |prop|
        valorProp = (prop.numCasas + prop.numHoteles + prop.precioCompra) * prop.precioEdificar
        capital += valorProp
            
        if (prop.hipotecada)
          capital -= prop.hipotecaBase
        end
      end
        return capital
    end
     
    def obtenerPropiedades(hipotecada)
      
      nuevo =Array.new
        
      @propiedades.each do |prop|
        if (prop.hipotecada() == hipotecada)
          nuevo << (prop)
        end
      end       
        
      return nuevo
    end    
    
    def pagarAlquiler()
      costeAlquiler = @casillaActual.pagarAlquiler()
      modificarSaldo(-costeAlquiler)
    end
    
    def pagarImpuesto()        
      @saldo -= @casillaActual.coste
    end
    
    def pagarLibertad(cantidad)
      tengoSaldo = tengoSaldo(cantidad)
        
        if (tengoSaldo)
            @encarcelado=(false)
            modificarSaldo(-cantidad);           
        end
    end
    
    
    
    def tengoCartaLibertad()
      return (@cartaLibertad != nil)
    end
    
    def tengoSaldo(cantidad)
      return (@saldo > cantidad) #superior o >= ?
    end
    
    
    def venderPropiedad(casilla)
      titulo = casilla.titulo
      eliminarDeMisPropiedades(titulo)
      precioVenta = titulo.calcularPrecioVenta()
      modificarSaldo(precioVenta)
        
    end
    
    private :eliminarDeMisPropiedades, :esDeMiPropiedad
    
      
    def <=>(otroJugador)
      otroJugador.obtenerCapital <=> obtenerCapital
                  
      otroCapital= otroJugador.obtenerCapital
      miCapital=obtenerCapital      
      if (otroCapital>miCapital)
        return 1 
      end
      if (otroCapital<miCapital)
        return -1 
      end      
      return 0
    end
    
    
    
    def stringpropiedades
      hey = ""
      @propiedades.each do |prop|      
        hey = " " + prop.to_s
      end     
      return hey
    end
      
      
    def to_s
      "nombre= #{@nombre} encarcelado=#{@encarcelado} \n saldo= #{@saldo} \n CAPITAL= #{obtenerCapital()} \n cartaLibertad= #{@cartaLibertad} \n casillaActual= #{@casillaActual} \n propiedades= #{stringpropiedades}"  
    end
    
    
    
    
    def convertirme(fianza)
      return Especulador.new(self, fianza)
    end
    
    def deboIrACarcel()
        return !tengoCartaLibertad()
    end
    
    def puedoEdificarCasa(titulo)
        edificada = false
        
        numCasas = titulo.numCasas
        if (numCasas < 4)
            costeEdificarCasa = titulo.precioEdificar
            tengoSaldo = tengoSaldo(costeEdificarCasa)
            
            if (tengoSaldo)
                titulo.edificarCasa()
                modificarSaldo(-costeEdificarCasa)
                edificada = true
            end            
        end
        
        return edificada
    end
    
    def puedoEdificarHotel(titulo)
        
        edificada = false
        
        numHoteles = titulo.numHoteles
        numCasas = titulo.numCasas
        if (numHoteles < 4 && numCasas >= 4)
            costeEdificarHotel = titulo.precioEdificar
            tengoSaldo = tengoSaldo(costeEdificarHotel)
            
            if (tengoSaldo)
                titulo.edificarHotel()
                modificarSaldo(-costeEdificarHotel)
                edificada = true
            end         
        end
        
        return edificada
    end
    
    
    #protected :convertirme, :deboIrACarcel, :puedoEdificarCasa, :puedoEdificarHotel, :tengoSaldo, :pagarImpuesto
    
    
    
    
  end
end

File: PFinalQytetetJava/QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;

/**
 *
 * @author antonio-w10
 */
public class Jugador implements Comparable<Jugador> {
    
    private boolean encarcelado = false;
    private String nombre;
    int saldo = 7500;
    
    private Sorpresa cartaLibertad;
    private Casilla casillaActual;
    private ArrayList<TituloPropiedad> propiedades;
    
    public Jugador(String nombre){
        this.nombre = nombre;
        this.cartaLibertad = null;
        this.casillaActual = null;
        this.propiedades = new ArrayList(); 
        
        
    }
    
    public Jugador(String nombre, Sorpresa cartaLibertad, Casilla casillaActual,ArrayList<TituloPropiedad> propiedades) {
        this.nombre = nombre;
        this.cartaLibertad = cartaLibertad;
        this.casillaActual = casillaActual;
        this.propiedades = propiedades;
        
        
    }
    
 
    boolean cancelarHipoteca(TituloPropiedad titulo){
        boolean cancelar = false;
        int costecancelar = titulo.calcularCosteCancelar();
        
        if (tengoSaldo(costecancelar)){
            
            modificarSaldo(-costecancelar);
            titulo.cancelarHipoteca();
            cancelar = true;
        }        
        
        return cancelar;
    }
     
    boolean comprarTituloPropiedad(){        
        boolean comprado = false;
        
        int costeCompra = this.casillaActual.getCoste();
        if (costeCompra < this.saldo){
            TituloPropiedad titulo = ((Calle)this.casillaActual).asignarPropietario(this);//jugactual;
            this.propiedades.add(titulo);
            modificarSaldo(-costeCompra);
            
            comprado = true;
            
        }
        return comprado;
    }
     
    int cuantasCasasHotelesTengo(){
         
        int num = 0;
        for (TituloPropiedad prop: this.propiedades){
            num += prop.getNumCasas() + prop.getNumHoteles();
        }
        return num;
    }
      
    boolean deboPagarAlquiler(){    
        TituloPropiedad titulo = ((Calle)this.casillaActual).getTitulo();
        
        boolean esDeMiPropiedad = esDeMiPropiedad(titulo);
        boolean tienePropietario = false;
        boolean encarcelado = true;
        boolean estaHipotecada = true;
        
        
        
        if (!esDeMiPropiedad){
            tienePropietario =  titulo.tengoPropietario();
            
            if(tienePropietario){
                encarcelado = titulo.propietarioEncarcelado();
            
                estaHipotecada = titulo.getHipotecada();
            }        
        }
                    
        return !esDeMiPropiedad && tienePropietario && !encarcelado && !estaHipotecada ;
    }
     
    Sorpresa devolverCartaLibertad(){
        Sorpresa cartaLibertad = this.cartaLibertad;
        this.cartaLibertad = null;
        return cartaLibertad;
    }
     
    boolean edificarCasa(TituloPropiedad titulo){
        return puedoEdificarCasa(titulo);
    }
     
    boolean edificarHotel(TituloPropiedad titulo){
        return puedoEdificarHotel(titulo);
    }
     
     
    private void eliminarDeMisPropiedades(TituloPropiedad titulo){
         this.propiedades.remove(titulo);
         titulo.setPropietario(null);
    }
     
    private boolean esDeMiPropiedad(TituloPropiedad titulo){
        return this.propiedades.contains(titulo);
    }
     
    boolean estoyEnCalleLibre(){
         throw new UnsupportedOperationException("Sin implementar");
    }
     

    boolean getEncarcelado() {
        return encarcelado;
    }

    String getNombre() {
        return nombre;
    }

    public int getSaldo() {
        return saldo;
    }

    Sorpresa getCartaLibertad() {
        return cartaLibertad;
    }

    Casilla getCasillaActual() {
        return casillaActual;
    }

    ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }
     
    
    void hipotecarPropiedad(TituloPropiedad titulo){
        int costeHipoteca = titulo.hipotecar();
        modificarSaldo(costeHipoteca);
        
    }
    
    void irACarcel(Casilla casilla){
        setCasillaActual(casilla);
        setEncarcelado(true);
    }
    
    int modificarSaldo(int cantidad){
        return this.saldo += cantidad;   
    }
    
    int obtenerCapital(){
        int capital = this.saldo;
        int valorProp = 0;
        
        for (TituloPropiedad prop: this.propiedades){
            valorProp = (prop.getNumCasas() + prop.getNumHoteles() + prop.getPrecioCompra()) * prop.getPrecioEdificar();
            capital += valorProp ;
            
            if (prop.getHipotecada()){
                capital -= prop.getHipotecaBase();
            }
        }
        return capital;
    }
     
    ArrayList<TituloPropiedad> obtenerPropiedades(boolean hipotecada){
        
        ArrayList<TituloPropiedad> nuevo = new ArrayList(); 
        
        for (TituloPropiedad prop: this.propiedades){
            if (prop.getHipotecada() == hipotecada){
                nuevo.add(prop);
            }
        }        
        
        return nuevo;    
    }
    
    
    void pagarAlquiler(){        
        int costeAlquiler = ((Calle)this.casillaActual).pagarAlquiler();
        modificarSaldo(-costeAlquiler);
        
    }
    
    protected void pagarImpuesto(){        
        this.saldo -= this.casillaActual.getCoste();
        
    }
    
    void pagarLibertad(int cantidad){
        boolean tengoSaldo = tengoSaldo(cantidad);
        
        if (tengoSaldo){
            setEncarcelado(false);
            modificarSaldo(-cantidad);            
        }
    }
    
    void setCartaLibertad(Sorpresa carta){
        this.cartaLibertad = carta;
    }
    
    void setCasillaActual(Casilla casilla){
        this.casillaActual = casilla;
    }
    
    void setEncarcelado(boolean encarcelado){
        this.encarcelado = encarcelado;
    }
    
    boolean tengoCartaLibertad(){
        return (this.cartaLibertad != null); 
    }
    
    protected boolean tengoSaldo(int cantidad){
       return (this.saldo > cantidad); //superior o >= ?
    }
    
    
    void venderPropiedad(Casilla casilla){
                
        TituloPropiedad titulo = ((Calle)casilla).getTitulo();
        eliminarDeMisPropiedades(titulo);
        int precioVenta = titulo.calcularPrecioVenta();
        modificarSaldo(precioVenta);
        
    }

    @Override
    public String toString() {
        return "Jugador{" + "encarcelado=" + encarcelado + ", nombre=" + nombre + ", saldo=" + saldo + ", CAPITAL=" + this.obtenerCapital() + ", cartaLibertad=" + cartaLibertad + ", casillaActual=" + casillaActual + ", propiedades=" + propiedades + '}';
    }
    
    
    @Override
    public int compareTo(Jugador otroJugador) {
        int otroCapital = ((Jugador) otroJugador).obtenerCapital();
        return (otroCapital - obtenerCapital());
        //descendente o ascendente cambiando la posicion del menos
        
    }
    
    
    protected Jugador(Jugador otroJugador){
        if(this != otroJugador){
            
            this.nombre = otroJugador.getNombre();
            this.cartaLibertad = otroJugador.getCartaLibertad();
            this.casillaActual = otroJugador.getCasillaActual();
            this.propiedades =otroJugador.getPropiedades();
            this.encarcelado = otroJugador.getEncarcelado();
            this.saldo = otroJugador.getSaldo();
        }
    }
    
    protected Especulador convertirme(int fianza){
        return new Especulador(this, fianza);
    }
    
    protected boolean deboIrACarcel(){
        return !tengoCartaLibertad();
    }
    
    protected boolean puedoEdificarCasa(TituloPropiedad titulo){
        boolean edificada = false;
        
        int numCasas = titulo.getNumCasas();
        if (numCasas < 4){
            int costeEdificarCasa = titulo.getPrecioEdificar();
            boolean tengoSaldo = tengoSaldo(costeEdificarCasa);
            
            if (tengoSaldo){
                titulo.edificarCasa();
                modificarSaldo(-costeEdificarCasa); 
                edificada = true;
            }            
        }
        
        return edificada;
    }
    
    protected boolean puedoEdificarHotel(TituloPropiedad titulo){
        // 
        //LO MISMO QUE CASA
        //
        boolean edificada = false;
        
        int numHoteles = titulo.getNumHoteles();
        int numCasas = titulo.getNumCasas();
        if (numHoteles < 4 && numCasas >= 4){
            int costeEdificarHotel = titulo.getPrecioEdificar();
            boolean tengoSaldo = tengoSaldo(costeEdificarHotel);
            
            if (tengoSaldo){
                titulo.edificarHotel();
                modificarSaldo(-costeEdificarHotel); 
                edificada = true;
            }            
        }
        
        return edificada;  
    }
    


    
}
